File: Behavioral/TemplateMethod/TemplateMethod.swift
// Implementation

class Computer {
    func printBrandName() {
        print("Computer brand: \(brandName())")
    }
 
    func brandName() -> String {
        assertionFailure("Please provide brand name")
        return ""
    }
 }

class Apple: Computer {
    override func brandName() -> String {
        return "Apple"
    }
}

class Dell: Computer {
    override func brandName() -> String {
        return "Dell"
    }
}

// Usage

let apple = Apple()
apple.printBrandName()

let dell = Dell()
dell.printBrandName()

File: Behavioral/TemplateMethod/TemplateMethod.kt
// Implementation

open class Computer {
    fun printBrandName() {
        println("Computer brand: " + brandName())
    }

    open fun brandName(): String {
        return "Default brand name"
    }
}

class Apple: Computer() {
    
    override fun brandName(): String {
        return "Apple"
    }
    
}

class Dell: Computer {
    override fun brandName(): String {
        return "Dell"
    }
}

// Usage

val apple = Apple()
apple.printBrandName()

val dell = Dell()
dell.printBrandName()